src/iterators/idlist_iterator.c

```diff
@@ -125,4 +119,81 @@ QueryIterator *IT_V2(NewIdListIterator) (t_docId *ids, t_offset num, double weig
   ret->Rewind = IL_Rewind;
 
   return ret;
-}
\ No newline at end of file
+}
+
+static void SetYield(QueryIterator *mr) {
+  ResultMetrics_Reset(mr->current);
+  ResultMetrics_Add(mr->current, NULL, RS_NumVal(mr->current->num.value));
+}
+
+static IteratorStatus MR_Read(QueryIterator *base) {
+  MetricIterator *mr = (MetricIterator *)base;
+  IdListIterator *it = &mr->base;
+  IteratorStatus rc = IL_Read(base);
+  if (ITERATOR_OK == rc) {
+    base->current->num.value = mr->metricList[it->offset - 1];
+    SetYield(base);
+  }
+  return rc;
+}
+
+static IteratorStatus MR_SkipTo(QueryIterator *base, t_docId docId) {
+  MetricIterator *mr = (MetricIterator *)base;
+  IdListIterator *it = &mr->base;
+  int rc = IL_SkipTo(base, docId);
+  if (ITERATOR_OK == rc || ITERATOR_NOTFOUND == rc) {
+    base->current->num.value = mr->metricList[it->offset - 1];
+    SetYield(base);
+  }
+  return rc;
+}
+
+static void MR_Free(QueryIterator *self) {
+  MetricIterator *mi = (MetricIterator *)self;
+  if (mi == NULL) {
+    return;
+  }
+  IdListIterator *it = &mi->base;
+  QueryIterator *base = &it->base;
+
+  IndexResult_Free(base->current);
+
+  if (it->docIds) {
+    rm_free(it->docIds);
+  }
+
+  if (mi->metricList) {
+    rm_free(mi->metricList);
+  }
```

Comment:
`free` (and `rm_free`) can handle `NULL` input. Consider removing the check (we also don't expect `NULL` values here)